backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -424,13 +436,71 @@ internal class ObjCExportNamerImpl(
     private fun String.startsWithWords(words: String) = this.startsWith(words) &&
             (this.length == words.length || !this[words.length].isLowerCase())
 
+    private inner class GenericTypeParameterNameMapping {
+        private val elementToName = mutableMapOf<TypeParameterDescriptor, String>()
+        private val typeParameterNameClassOverrides = mutableMapOf<ClassDescriptor, MutableSet<String>>()
+
+        fun reserved(name: String): Boolean {
+            return name in reservedNames
+        }
+
+        fun getOrPut(element: TypeParameterDescriptor, nameCandidates: () -> Sequence<String>): String {
+            getIfAssigned(element)?.let { return it }
+
+            nameCandidates().forEach {
+                if (tryAssign(element, it)) {
+                    return it
+                }
+            }
+
+            error("name candidates run out")
+        }
+
+        private fun tryAssign(element: TypeParameterDescriptor, name: String): Boolean {
+            if (element in elementToName) error(element)
+
+            if (reserved(name)) return false
+
+            if (!validName(element, name)) return false
+
+            assignName(element, name)
+
+            return true
+        }
+
+        private fun assignName(element: TypeParameterDescriptor, name: String) {
+            if (!local) {
+                elementToName[element] = name
+            }
+            classNameSet(element).add(name)
+        }
+
+        private fun validName(element: TypeParameterDescriptor, name: String): Boolean {
+            assert(element.containingDeclaration is ClassDescriptor)
+
+            val nameSet = classNameSet(element)
+            return !objCClassNames.nameExists(name) && !objCProtocolNames.nameExists(name) && name !in nameSet
+        }
+
+        private fun classNameSet(element: TypeParameterDescriptor): MutableSet<String> {
+            return typeParameterNameClassOverrides.getOrPut(element.containingDeclaration as ClassDescriptor) {
```

Comment:
Doesn't seem correct for (yet unsupported) outer class type parameters.
```
class Foo<T> {
    inner class Bar<T>
}
```

Reply:
Generics are definitely not supporting outer generics on nested or inner types at this point. Although the resulting syntax isn't great, it's possible to cast your way around it. Can be added, though. Will defer to you. Will add if you think this would be valuable.